### PR TITLE
Feture and target extraction pipelines

### PR DESCRIPTION
**Pull request by jano1906**

### CODE REVIEW

File: jointformer/models/transformer.py [hunk 1]
```diff
@@ -60,12 +57,20 @@ def forward(
         x = self.transformer.drop(tok_emb + pos_emb)
         attention_probs = []
         for block in self.transformer.h:
-            x, attn = block(x, is_causal=is_causal, mask=attention_mask)
+            x, attn = block(x, is_causal=is_causal, mask=None if is_causal else attention_mask)
             attn = attn.detach().cpu() if attn is not None else None
             attention_probs.append(attn)
         x = self.transformer.ln_f(x)
 
+        if attention_mask is None:
```
Comment (AdamIzdebski) [inline jointformer/models/transformer.py:65]: Czy nie chodzilo nam o to?
```
global_embedding = torch.einsum('bsd, bs -> bsd', x, attention_mask.float()) if attention_mask is not None else x
global_embedding = global_embedding.mean(-1)    
```

File: experiments/data/extract_features.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    parser.add_argument("--path_to_model_config", type=str, required=True)
+    parser.add_argument("--path_to_task_config", type=str, required=True)
+    parser.add_argument("--output", type=str, required=True)
+    parser.add_argument("--batch_size", type=int, default=64)
+    parser.add_argument("--chunk_size", type=int, default=100000)
+    parser.add_argument("--device", type=str, default="cuda:0")
+    return parser
+
+
+def extract_features(model: SmilesEncoder, data: list[str], output_path: str, chunk_size: int):
```
[the remark this answers is not in the document]
Comment (jano1906) [reply experiments/data/extract_features.py:28]: An instance of SmilesEncoder is provided with tokenizer at creation so it should be able to process any string from a dataset. By default an instance of BaseDataset applies transforms to input but if those were to be applied it would be implemented in tokenizer instead since input transform is model dependent, not dataset dependent. As for now I don't think there is a reason to abstract.

File: jointformer/models/transformer.py [hunk 3]
```diff
@@ -60,12 +57,20 @@ def forward(
         x = self.transformer.drop(tok_emb + pos_emb)
         attention_probs = []
         for block in self.transformer.h:
-            x, attn = block(x, is_causal=is_causal, mask=attention_mask)
+            x, attn = block(x, is_causal=is_causal, mask=None if is_causal else attention_mask)
```
[the remark this answers is not in the document]
Comment (jano1906) [reply jointformer/models/transformer.py:60]: Beware it doesn't fix the bug with hardcoding causal masking. This just allows us to use attention mask later in the computation of global embedding. Before the attention mask was overridden to None.

File: jointformer/models/transformer.py (hunk 1 above)
Comment (jano1906) [reply jointformer/models/transformer.py:65]: 1. I think you meant
```
global_embedding = global_embedding.mean(-2)   
```
in your implementation. We should not average over representation dimension but over the sequence length. 
2. We want to take the mean over the sequence and sequences have variable length. This is why we normalize the attention mask and do matmul.

File: experiments/data/extract_features.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AdamIzdebski) [reply experiments/data/extract_features.py:28]: ok

File: jointformer/models/transformer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AdamIzdebski) [reply jointformer/models/transformer.py:60]: yes, I just agree with explicitely passing a None when is_causal is True

File: jointformer/models/transformer.py (hunk 1 above)
Comment (AdamIzdebski) [reply jointformer/models/transformer.py:65]: Sorry for the confusion - yes, I meant 
global_embedding = global_embedding.mean(-2)

Comment (AdamIzdebski) [reply jointformer/models/transformer.py:65]: ok, I see now

**Review by AdamIzdebski — APPROVED** [review]
